### ISSUE TITLE
Ambiguity introduced by the namespace change

### ISSUE THREAD

**Issue by chrrasmussen** [idris-lang/Idris2#632]
It seems like #631 has introduced some new ambiguity related to the List and Unit types. At the bottom I have included a set of changes that I could apply to resolve the compile errors. However, the error related to `writeToTTC` (in `src/Yaffle/Main.idr`) seems like something that could occur in the wild.

# Steps to Reproduce

```
git clone https://github.com/idris-lang/Idris2
cd Idris2
make bootstrap SCHEME=chez
make install

make clean
make all
make install

make clean
make all
```

Notice that I am building the compiler one extra time, compared to what the CI does.

# Expected Behavior

Type-checks.

# Observed Behavior

I get compile error in the following files (Click to expand):

<details>
  <summary>src/Text/Bounded.idr</summary>

```
14/158: Building Text.Bounded (src/Text/Bounded.idr)
Error: While processing right hand side of showPrec. Ambiguous elaboration. Possible results:
    ?postpone [locals in scope: ty, {conArg:446}, val, ec, el, sc, sl, ir, d]
    ?arg Prelude.:: ?arg

src/Text/Bounded.idr:31:35--31:108
    |
 31 |     showCon d "MkBounded" (concat [showArg ir, showArg val, showArg sl, showArg sc, showArg el, showArg ec])
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
</details>

<details>
  <summary>src/TTImp/Elab/RunElab.idr</summary>

```
98/158: Building TTImp.Elab.RunElab (src/TTImp/Elab/RunElab.idr)
Error: While processing right hand side of elabScript. While processing right hand side of elabScript,elabCon. Can't find an
implementation for Reflect Type.

src/TTImp/Elab/RunElab.idr:72:14--72:26
    |
 72 |              scriptRet ()
    |              ^^^^^^^^^^^^
```
</details>

<details>
  <summary>src/Yaffle/Main.idr</summary>
  
```
156/158: Building Yaffle.Main (src/Yaffle/Main.idr)
Error: While processing right hand side of yaffleMain. Can't find an implementation for HasNames Type.

src/Yaffle/Main.idr:65:29--65:72
    |
 65 |                             writeToTTC () !(getTTCFileName fname "ttc")
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
</details>

# Workaround

```diff
diff --git a/src/TTImp/Elab/RunElab.idr b/src/TTImp/Elab/RunElab.idr
index 3b1c99e7..63e6e8dd 100644
--- a/src/TTImp/Elab/RunElab.idr
+++ b/src/TTImp/Elab/RunElab.idr
@@ -69,7 +69,7 @@ elabScript fc nest env (NDCon nfc nm t ar args) exp
              logC !(reify defs topic') !(reify defs verb') $
                   do str' <- evalClosure defs str
                      reify defs str'
-             scriptRet ()
+             scriptRet (the Unit ())
     elabCon defs "LogTerm" [topic, verb, str, tm]
         = do topic' <- evalClosure defs topic
              verb' <- evalClosure defs verb
@@ -78,7 +78,7 @@ elabScript fc nest env (NDCon nfc nm t ar args) exp
                      tm' <- evalClosure defs tm
                      pure $ !(reify defs str') ++ ": " ++
                              show (the RawImp !(reify defs tm'))
-             scriptRet ()
+             scriptRet (the Unit ())
     elabCon defs "Check" [exp, ttimp]
         = do exp' <- evalClosure defs exp
              ttimp' <- evalClosure defs ttimp
@@ -159,7 +159,7 @@ elabScript fc nest env (NDCon nfc nm t ar args) exp
         = do d' <- evalClosure defs d
              decls <- reify defs d'
              traverse_ (processDecl [] (MkNested []) []) decls
-             scriptRet ()
+             scriptRet (the Unit ())
     elabCon defs n args = failWith defs
 elabScript fc nest env script exp
     = do defs <- get Ctxt
diff --git a/src/Text/Bounded.idr b/src/Text/Bounded.idr
index 79ff3d58..6bbe0341 100644
--- a/src/Text/Bounded.idr
+++ b/src/Text/Bounded.idr
@@ -28,7 +28,7 @@ Eq ty => Eq (WithBounds ty) where
 export
 Show ty => Show (WithBounds ty) where
   showPrec d (MkBounded val ir sl sc el ec) =
-    showCon d "MkBounded" (concat [showArg ir, showArg val, showArg sl, showArg sc, showArg el, showArg ec])
+    showCon d "MkBounded" (concat {t = List} [showArg ir, showArg val, showArg sl, showArg sc, showArg el, showArg ec])
 
 export
 Functor WithBounds where
diff --git a/src/Yaffle/Main.idr b/src/Yaffle/Main.idr
index 22dc0769..5eb14490 100644
--- a/src/Yaffle/Main.idr
+++ b/src/Yaffle/Main.idr
@@ -62,7 +62,7 @@ yaffleMain fname args
                       when ok $
                          do ns <- pathToNS (working_dir d) (source_dir d) fname
                             makeBuildDirectory ns
-                            writeToTTC () !(getTTCFileName fname "ttc")
+                            writeToTTC (the Unit ()) !(getTTCFileName fname "ttc")
                             coreLift $ putStrLn "Written TTC"
          ust <- get UST
```

**Comment by chrrasmussen** [comment]
I figured I should add a smaller reproducer:

```idris
foo : Semigroup a => a -> ()
foo x = ()

main : IO ()
main = pure (foo ())

{-
Error: While processing right hand side of main. Can't find an implementation for Semigroup Type.

Test.idr:8:14--8:20
   |
 8 | main = pure (foo ())
   |   
-}
```

As above, changing the code to `main = pure (foo (the Unit ()))` works.

**Comment by gallais** [comment]
I had to fix some ambiguities in one of the reflection test cases.
Something of the form `traverse f [1,2,3]`. The problem being that
`[1,2,3]` as a literal could be a `List` or a `List1` and `traverse`
works for both. The result was being discarded so there was no hope
of later constraints solving this ambiguity.

The `concat` problem is similar: it is inherently ambiguous to have
a function of type `t a -> a` applied to a literal: there is no reliable
way to infer what `t` ought to be!

**Comment by gallais** [comment]
Wait. How come it's showing up after the *second* build?
After the bootstrap we are already building with the new
compiler and libraries, aren't we? :scream:

**Comment by gallais** [comment]
Looks like the problem was assuming that `Namespace`s are necessarily non-empty.

**Comment by chrrasmussen** [comment]
@gallais fixed this in #638. Thank you! ❤️
